keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -323,11 +338,180 @@ myOtherField: newly added field
+ provide a minimum level of ratcheting support for any validation rule in the 
+ schema.
+
+`DeepEqual(OldObject, NewObject)` requirement is very strong. But for some 
+validation rules it is too strict. Consider the `required` error. This
+validation is applied to the object owning the property. Thus, if a new field
+becomes required the object cannot have other fields added onto it or changed
+without also adding the newly required field.
+
+Kubernetes has precedent for using a weaker form of racheting when validation is
+strengthened:
+
+```go
+if !Validation(old) {
+  // Ignore errors caused by !Validation(new)
+}
+```
+
+An example of this is the structural schema requirement added when Kubernetes
+went GA ([source](https://github.com/kubernetes/kubernetes/blob/12dc19d46fb97cbbfeb1e12b8a10ff7ae73d9515/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1583-L1590)): 
+```go
+// requireStructuralSchema returns true if schemas specified must be structural
+func requireStructuralSchema(oldCRDSpec *apiextensions.CustomResourceDefinitionSpec) bool {
+	if oldCRDSpec != nil && specHasNonStructuralSchema(oldCRDSpec) {
+		// don't tighten validation on existing persisted data
+		return false
+	}
+	return true
+}
+```
+
+It is not possible to generically allow this weaker form of ratcheting on all
+types of validations - it may allow data that would have failed a validation
+on an older schema.
+
+To enable schemas to add validations with more complicated ratcheting logic
+than a DeepEqual check (e.g. to allow a new `minLength` restriction to allow any
+value as long as it is getting closer the new minimum); we turn to CEL.
+
+#### Ratcheting Rules in CEL
+
+CRD Validation Rules currently have two types:
+
+- Normal Rule (Applies to `CREATE` and `UPDATE` requests)
+- Transition Rule (Applies to `UPDATE` requests only)
+
+With only these rule types we are not able to express any logic of the form 
+"old values must obey one schema, and new values another".
+
+This is way we propose to add a new rule type to enable this functionality for
+ratcheting:
+
+- Create-Only Rule (Applies to `CREATE` requests)
+
+This would involve the following change to the `ValidationRule` type:
+
+```go
+// ValidationRule describes a validation rule written in the CEL expression language.
+type ValidationRule struct {
+  // ...
+	Rule string `json:"rule" protobuf:"bytes,1,opt,name=rule"`
+	Message string `json:"message,omitempty" protobuf:"bytes,2,opt,name=message"`
+	MessageExpression string `json:"messageExpression,omitempty" protobuf:"bytes,3,opt,name=messageExpression"`
+	Reason *FieldValueErrorReason `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
+	FieldPath string `json:"fieldPath,omitempty" protobuf:"bytes,5,opt,name=fieldPath"`
+  // ...
+
+  // New
+  // Request types that this validation rule applies to
+  // +default=[Create,Update], or just [Update] if `oldSelf` is used
```

Comment:
There's are API nuances that will make this default a problem in practice-

If we use defaulting, and `oldSelf` is in the CEL expression, you'll get the correct default when you do a CREATE. But if you then do a subsequent UPDATE and remove `oldSelf`, the default won't change.  This is backward incompatible with how the fields work today since you can remove `oldSelf` in an expression today and the operations will automatically and implicitly change to CREATE and UPDATE.

Reply:
Thanks for pointing this out. Can we work around this by saying the field is not defaulted, but the behavior would be that if it is unset the rule applies wherever applicable? I've updated the field comment to replace `+default` with this

Reply:
An alternative to that would be to default to `OperationAll` which means the same thing - all applicable expressions

Reply:
As much as I like the idea of defaulting the value to what the actual behavior will be-- because it's great API feedback. I think we should not default this field and instead make it optional and use the behavior we have today if they are unset.

We might need to grid out the behavior for when `oldSelf`  is present/absent for all possible operations values.  If operations=CREATE and oldSelf is used, what happens? Do we just not allow that combination, or do we set oldSelf to be null?

Reply:
> If operations=CREATE and oldSelf is used, what happens?

I was thinking a validation error would be thrown

Reply:
> I was thinking a validation error would be thrown

For native types, that's probably right.

For CRDs, we COULD fail CRD validation when the CRD author attempts to CREATE or UPDATE a CRD with a schema that includes this, since it's invalid and serves no possible purpose.

Reply:
Sorry, thats what I meant by validation error is thrown - a validation error of the CRD itself